### ISSUE TITLE
Exported functions are flagged as 'unused'

### ISSUE THREAD

**Issue by ianmackenzie** [klazuka/intellij-elm#325]
If a function is defined in a file and exported but not otherwise used in that file, it gets flagged with a warning as 'unused'. If a function/value is exported, shouldn't that count as it being 'used'?

**Comment by klazuka** [comment]
It could go either way.

For *package* projects where the module is listed in the `exposed-modules` section in your `elm.json`, I think you're right, it should be treated as used.

But for all other cases, I think it's better if it is marked as unused so that you know about it and determine if that is a mistake or not.

@ianmackenzie are you working on an application or a package?

**Comment by ianmackenzie** [comment]
In this case I was indeed working on a package - but wouldn't the same logic apply to a module inside an app that provided some useful functionally to the rest of the app? If I had a `User` module in my app with `firstName`, `lastName` and `fullName` functions, it seems reasonable that (especially) `fullName` might not be used within the `User` module but *would* be used elsewhere in the app.

The gold standard would likely be "is this ultimately reachable by an exposed function in an exposed module" (for a package) or "is this reachable from `main`" (for an app) but I'm guessing that kind of whole-program analysis could get expensive.

**Comment by ianmackenzie** [comment]
(Sorry, fat fingers on mobile...didn't mean to close the issue!)

**Comment by klazuka** [comment]
I don’t expect that we will be doing control flow analysis any time soon (if ever). And even if we did, it’d probably be too expensive to run for every symbol in the file. 

The existing situation for application projects is fine, although there may be false negatives. 

I will fix the case for packages.

**Comment by ianmackenzie** [comment]
Sorry, I should have chosen better terminology...I wasn't thinking actual control flow analysis, more just checking if there was any chain of function references that led from `main` (or an exposed function of an exposed module) to the individual function in question (without trying to check if any of those references were inside an `if False` or something).

**Comment by klazuka** [comment]
Your idea would be a good middle-ground, but it still ends up being kinda expensive. Right now the “unused symbol” inspection runs in the background for every named thing in the current file. So every time you type a key, it has to re-run the inspection, and it’s expensive to go out and fetch all the callers, transitively... and that has to be done for every symbol in the file.

My goal with this feature is to provide a very good approximation that will find most dead code. A false positive (like the example you gave at the top of this issue for a function that is exposed by an exposed module in a package) is a bug, and should/will be fixed. False negatives, on the other hand, are acceptable (at least for now when there are a lot of other features remaining to be implemented). 

Tools like elm-xref, elm-analyze and the Elm compiler itself are in a better position to do a more thorough analysis since they do not run interactively, and they are designed to run top-down (most algorithms used in IntelliJ are bottom-up, based on the current file, for efficiency reasons).

**Comment by ianmackenzie** [comment]
Thanks for the detailed explanation, makes sense!

Although it occurred to me that flagging exported but not otherwise used functions could be considered a _feature_...those would be good functions to add tests for first! Since they're not used by anything else in the module, they would never get tested indirectly, and conversely if you test all those "top level" functions then you have a good shot at indirectly testing everything else.

**Comment by klazuka** [comment]
Yeah, I was starting to think the same thing the other night. Another possibility here would be to change the error message so that it gives the user a little more guidance.

**Comment by viktor-ferenczi** [comment]
Code sample to reproduce:

```
module SomeModule exposing (something)

something =
    "something"
```

something will be grayed as unused, while it is exposed as public API of the module.
